tests/System.Text.Primitives.Tests/Encoding/Utf8EncoderTests.cs

```diff
@@ -4,6 +4,522 @@ namespace System.Text.Utf8.Tests
 {
     public class Utf8EncoderTests
     {
+        private const ushort Utf16HighSurrogateFirstCodePoint = 0xD800;
+        private const ushort Utf16HighSurrogateLastCodePoint = 0xDBFF;
+        private const ushort Utf16LowSurrogateFirstCodePoint = 0xDC00;
+        private const ushort Utf16LowSurrogateLastCodePoint = 0xDFFF;
+
+        private const byte Utf8OneByteLastCodePoint = 0x7F;
+        private const ushort Utf8TwoBytesLastCodePoint = 0x7FF;
+        private const ushort Utf8ThreeBytesLastCodePoint = 0xFFFF;
+
+        private const int CharLength = 1000;
+
+        [Fact]
+        public void TestEncodingInputBufferEmpty()
+        {
+            string unicodeString = "";
+            ReadOnlySpan<char> characters = unicodeString.Slice();
+            int expectedBytesWritten = GetByteCount(characters);
+            char[] charArray = characters.ToArray();
+            byte[] utf8BufferExpected = new byte[Encoding.UTF8.GetByteCount(charArray)];
+
+            Encoding.UTF8.GetBytes(charArray, 0, characters.Length, utf8BufferExpected, 0);
+            byte[] utf8Buffer = new byte[expectedBytesWritten + 25];
+            Span<byte> span = new Span<byte>(utf8Buffer);
+            int charactersConsumed;
+            int bytesWritten;
+            Assert.True(TextEncoder.Utf8.TryEncode(characters, span, out charactersConsumed, out bytesWritten));
+
+            Assert.Equal(0, charactersConsumed);
+            Assert.Equal(expectedBytesWritten, bytesWritten);
+        }
+
+        [Fact]
+        public void TestEncodingOutputBufferEmpty()
+        {
+            string unicodeString = GenerateValidString(CharLength, 0, Utf8ThreeBytesLastCodePoint);
+            ReadOnlySpan<char> characters = unicodeString.Slice();
+            char[] charArray = characters.ToArray();
+            byte[] utf8BufferExpected = new byte[Encoding.UTF8.GetByteCount(charArray)];
+
+            Encoding.UTF8.GetBytes(charArray, 0, characters.Length, utf8BufferExpected, 0);
+            byte[] utf8Buffer = new byte[0];
+            Span<byte> span = new Span<byte>(utf8Buffer);
+            int charactersConsumed;
+            int bytesWritten;
+            Assert.False(TextEncoder.Utf8.TryEncode(characters, span, out charactersConsumed, out bytesWritten));
+
+            Assert.Equal(0, charactersConsumed);
+            Assert.Equal(0, bytesWritten);
+        }
+
+        [Fact]
+        public void TestEncodingInputBufferMaxSize()
+        {
+            int maxCharArraySize = int.MaxValue / 4; // TODO: want this to be int.MaxValue (need gcAllowVeryLargeObjects = true)
+            char[] charArray = new char[maxCharArraySize];
+
```

Comment:
How can I get a int.MaxValue sized array so I can make sure there aren't Integer OverFlow issues resulting in incorrect encoding?
Right now CLR throws an OutofMemoryException for anything over 2GB.